extensions/eclipse-che-theia-remote/src/node/plugin-remote.ts

```diff
@@ -0,0 +1,225 @@
+    }
+
+    public getIdentifier(): number {
+        return this.id;
+    }
+
+    // message is a JSON entry
+    // tslint:disable-next-line:no-any
+    send(message: any) {
+        this.socket.send(JSON.stringify(message));
+    }
+
+}
+
+// list of clients
+const webSocketClients = new Map<number, WebSocketClient>();
+
+// create a new client on top of socket
+function newClient(id: number, socket: ws): WebSocketClient {
+    const webSocketClient = new WebSocketClient(id, socket);
+    const emitter = new Emitter();
+    webSocketClient.emitter = emitter;
+    webSocketClient.rpc = new RPCProtocolImpl({
+        onMessage: emitter.event,
+        // send messages to this client
+        send: (m: {}) => {
+            webSocketClient.send(m);
+        }
+    });
+
+    const pluginHostRPC = new PluginHostRPC(webSocketClient.rpc);
+    pluginHostRPC.initialize();
+
+    // override window.createTerminal to be container aware
+    // tslint:disable-next-line:no-any
+    new TerminalContainerAware().overrideTerminal((webSocketClient.rpc as any).locals[MAIN_RPC_CONTEXT.TERMINAL_EXT.id]);
+
+    return webSocketClient;
+}
+
+// Handle the connection received
+function handleConnection(socket: ws, request: http.IncomingMessage): void {
+    // create channel for discussing with this new client
+    const channelId = sessionId++;
+    const client = newClient(channelId, socket);
+    webSocketClients.set(channelId, client);
+
+    socket.on('error', err => {
+    });
+
+    socket.on('close', (code, reason) => {
+        webSocketClients.clear();
+    });
+
+    socket.on('message', (data: ws.Data) => {
+        const jsonParsed = JSON.parse(data.toString());
+
+        // handle local call
+        if (jsonParsed.internal) {
+            // asked to grab metadata, send them
+            if (jsonParsed.internal.metadata && 'request' === jsonParsed.internal.metadata) {
+                const metadataResult = {
+                    'internal': {
+                        'endpointName': jsonParsed.internal.endpointName,
+                        'metadata': {
+                            'result': currentBackendPluginsMetadata
+                        }
+                    }
+                };
+
+                client.send(metadataResult);
+            }
+            return;
+        }
+
+        for (const webSocketClient of webSocketClients.values()) {
+            // send what is inside the message (wrapped message)
+            try {
+                webSocketClient.emitter.fire(jsonParsed);
+            } catch (e) {
+                console.error(e);
+            }
+        }
+    });
+}
+
+const currentBackendPluginsMetadata: PluginMetadata[] = [];
+
+@injectable()
+class PluginDeployerHandlerImpl implements PluginDeployerHandler {
+
+    @inject(ILogger)
+    protected readonly logger: ILogger;
+
+    constructor(
+        @inject(HostedPluginReader) private readonly reader: HostedPluginReader,
```

Comment:
The same as with `HostedPluginMapping`

Reply:
I'm not sure what you want to say ?
it's an import
```
 import { HostedPluginReader } from '@theia/plugin-ext/lib/hosted/node/plugin-reader';
```

Reply:
Yes, I see. My concern was about using code which had been originally designed to work with plug-in under development: https://github.com/theia-ide/theia/blob/master/packages/plugin-ext/src/hosted/node/plugin-reader.ts#L49
We may reuse this code for remote plug-ins, but I feel that there is a mess in `hosted` package and naming... At least it seems to me so...